### ISSUE TITLE
[Feature request] Tell the user (and maybe a server) if the ROM is unknown

### ISSUE THREAD

**Issue by mariomadproductions** [melonDS-emu/melonDS#1657]
I think it might be useful for DS game preservation if melonds checked if the opened ROM's gamecode and revision were one that isn't in the internal database, and do some checks to see if the ROM looks genuine (not a mod or homebrew). It would then display a notice (in status bar or in-screen HUD I guess) saying its an undocumented ROM, and maybe send metadata to a server. That way more unpreserved games could be disovered.

**Comment by patataofcourse** [comment]
this feels a bit invasive, and also, there's really not an easy way to tell apart bad dumps or mods, especially mods that move around filesystem stuff

**Comment by Arisotura** [comment]
I am also opposed to any sort of telemetry

**Comment by nadiaholmquist** [comment]
Maybe this could work by including a no-intro database of valid DS dumps and warning the user if the dump they have is not in it.

Tell them something about how they might have a bad dump, or to reach out to preservation groups if they dumped a unique cart. This could get very annoying for ROM hacks though.

**Comment by mariomadproductions** [comment]
Yeah, I think there could just be an option to send info to the server (not an automatic thing), or it would just suggest to the user to get in contact.

For games using the TWL ROM format (ones from 2009 onwards, including normal DS games), the whole ROM is signed by Nintendo, so they can be verified as genuine that way.

For ones using the NTR ROM format (ones from before 2009), we would need a way to rule out mods, yeah. Not entirely sure how that could be done.

**Comment by mariomadproductions** [comment]
Although here are some things that may give a clue if an NTR ROM is a mod:
* Wrong checksums in header
* Wrong size in header
* More differences to the known ROM than just an underdumped header or a few corrupted bytes would cause, but less differences than would be caused by adding/replacing new files and code to the ROM.

**Comment by patataofcourse** [comment]
honestly that just feels like overkill

**Comment by mariomadproductions** [comment]
> honestly that just feels like overkill

Overkill for what?

**Comment by patataofcourse** [comment]
I think that, while the intentions for this are good, storing so many hashes of parts of every single dump of every single released (or unreleased) game for the NDS is just far too much effort and data for something like this.

A database like that would also need to be *managed*, and I don't think anyone in the melonDS team has to want to collect so many hashes for every single dump in the no-intro dataset

**Comment by patataofcourse** [comment]
hey, please don't delete your comment while i'm replying :P

**Comment by mariomadproductions** [comment]
> I think that, while the intentions for this are good, storing so many hashes of parts of every single dump of every single released (or unreleased) game for the NDS is just far too much effort and data for something like this.
> 
> A database like that would also need to be _managed_, and I don't think anyone in the melonDS team has to want to collect so many hashes for every single dump in the no-intro dataset

I can easily create that and update it.

> hey, please don't delete your comment while i'm replying :P

Sorry, I thought maybe my comment was unnecessary.

**Comment by mariomadproductions** [comment]
Myria just pointed out to me that NTR ROMs have header/banner/main executable/overlay hashes in the DSi/3DS cart whitelists. So that data could be parsed out and included in MelonDS. That means none of this "guessing" would be needed.

**Comment by nathanielcwm** [comment]
Dolphin implements it as a tab in the game properties menu that you have to explicitly seek for it to test your iso. I'm opposed to an automatic test on every rom especially if it is linked to telemetry.

![image](https://user-images.githubusercontent.com/22572406/231341250-89e61906-a2c0-4327-a593-445a63db4f52.png)

**Comment by CasualPokePlayer** [comment]
To be fair, Dolphin wouldn't do it automatically just due to how long it takes to hash an entire disc file (especially when it was first introduced). That situation isn't completely comparable with melonDS (especially as the entire file is loaded into memory at the moment, file I/O woud be one of the main slowdowns when hashing in the Dolphin example)

**Comment by mariomadproductions** [comment]
> Dolphin implements it as a tab in the game properties menu that you have to explicitly seek for it to test your iso. I'm opposed to an automatic test on every rom especially if it is linked to telemetry.
> 
> ![image](https://user-images.githubusercontent.com/22572406/231341250-89e61906-a2c0-4327-a593-445a63db4f52.png)

Dolphin's thing isn't exactly the same, as it doesn't have an internal database of known ROMs, so can't tell the user if they have something new. But yeah the thing where it checks the internal ROM hashes is similar.

I agree that melonds shouldn't make network requests without the user telling it to.
